### PR TITLE
Fix and extend `wcs1d-fits` loader for multi-D WCS

### LINKED ISSUE #592: Writer needed for cube data
How do we save spectral cubes to file formats that are readable by other applications?
What type of api call do we implement for this?

[🐱](https://jira.stsci.edu/browse/JDAT-2199)

### PR DESCRIPTION
**Pull request by dhomeier**
To address https://github.com/astropy/specutils/issues/592#issuecomment-1428846666; quick fix for enabling writing of spectral cubes with celestial components in WCS.
Reading in with the full WCS is now supported as well, plus adding `mask` and `uncertainty` to the attributes read from file.
The latter at this point are still limited to the format storing them all in different ImageHDUs; an alternative common format storing the fields as sub-arrays of an N+1 image yet to be implemented.
Fixes #592

### CODE REVIEW

**Comment by pllim** [comment]
So this fixes #592 ?

**Comment by dhomeier** [comment]
This should work for your example.
Reading the full WCS in rather than only the spectral dimension may of course have consequences if `spectrum.wcs` is used at a later point. Another immediate change is that this will now only accept the official FITS spectral axis keywords recognised my WCSLIB, `FREQ`, `AFRQ`, `ENER`, `WAVN`,`VRAD`, `WAVE`, `VOPT`, `ZOPT`, `AWAV`, `VELO`, `BETA`.
The tests are adapted, but if downstream applications have used names like `wavelength` or `wavenumber`, this may need additional pre-processing of the header.

**Comment by dhomeier** [comment]
I was somewhat surprised when updating the writer, that there are apparently WCS that have a `spectral` component, yet
`wcs.spectral.all_pix2world(pix)` returns a different (wrong) result than `wcs.all_pix2world(pix)` (the former fails on all other tests that only input a 1D, pure spectral WCS). But the check should probably be there anyway, if it ever has to deal with non-FITS WCS.

File: specutils/tests/test_loaders.py
Comment (nmearl) [inline specutils/tests/test_loaders.py:710]: Just some minor context: if I recall, the spectral axis names were meant to correspond with the class properties and not the FITS naming standard. Either way, it's not an issue, but figured the history of the usage might add some context.

**Comment by dhomeier** [comment]
I have added read and write for `spectrum.mask` and given the HDUs names, handling it I think more similar to how the cubeviz loader treats its files. Things like detection strategy for the spectrum HDU (and mask? Should it be found under other names as well?) may need some more discussion, also perhaps more test coverage for the naming schemes.
In principle this could probably extended in a similar way to also handle uncertainties.

**Comment by dhomeier** [comment]
Apparently something in oldestdeps is pulling numpy 1.24.2 in, overriding `numpy==1.19` and breaking the old scipy version. I guess that means 1.19 is not really supported anymore, but don't know what it is that's requiring a newer version.

**Comment by dhomeier** [comment]
Marking this as ready for review now, as the multi-HDU format for masks and uncertainties is functional (unless there are more labels the loader needs to identify).
The format reading different attributes from `data` slices in a single HDU, which would work similarly to 6dfgs
https://github.com/astropy/specutils/blob/6eb7f96498072882c97763d4cd10e07cf81b6d33/specutils/io/default_loaders/sixdfgs_reader.py#L188-L189
could still be added here or in a follow-up PR.

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:48]: Indentation should match inner parenthesis.
```suggestion
                 hdulist[hdu].header.get('WCSAXES', 0) == 1 or
                 hdulist[hdu].header.get('DISPAXIS', -1) >= 0 ) and not
```

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:81]: This confuses me. Is this loader only for *image*? A spectrum can also come as table HDU, right?

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:141]: Why would mask be represented as `str`?

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:226]: Unless `spectrum.flux.unit` actually generates API doc, use double-backticks. I cannot check because RTD build failed on your PR. Also see #1028.

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:240]: This comment does not inspire confidence. Can someone well-versed with spectrum WCS chime in here?

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:265]: Usually mask only makes sense in pixel space. Putting WCS on that extension would be misleading. In HST images, I think the WCS is only in the SCI extensions.

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:279]: Should we even let user rename the uncertainty HDU? Would be misleading if `spectrum.uncertainty` is STD but then someone names it "IVAR".

File: specutils/tests/test_loaders.py
Comment (pllim) [inline specutils/tests/test_loaders.py:797]: Isn't it simpler to start with all zeroes and then assign select slices to ones? Is a random number generator really necessary? Looks cool though.

**Comment by pllim** [comment]
Thanks for working on this! I left some comments above.

Also, does this page need updating? It is not very clear to me exactly which data will go through this reader/writer.

https://specutils.readthedocs.io/en/latest/spectrum1d.html#list-of-loaders

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:81]: wcs1d handles only spectra with flux represented as image and spectral axis defined by the WCS; tables are covered by `tabular-fits`. That already supports masks and uncertainties, more or less, which are simply additional table columns in that representation.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:141]: `mask_hdu` is the generalised index to its HDU, which for a HDUList can be the numerical index or the label from `EXTNAME`.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:226]: [`spectrum.flux`](https://specutils.readthedocs.io/en/stable/api/specutils.Spectrum1D.html#specutils.Spectrum1D.flux) does, but not specifically `unit`, is that what you mean? I don't know if the loaders themselves generate API docs, or rather where they might be linked from ;-)
Thanks for pointing out the RTD failure; throws in all current PRs and does not seem related – from a quick glance some parsing error related to `reproject`, but needs further investigation.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:240]: The comment refers to my expectation for a WCS with a single, spectral dimension, that has a `spectral` (and thus also a `celestial`) attribute, to return the same world coordinates via that `wcs.spectral` as from `wcs` directly.
But that case is also handled separately in
https://github.com/astropy/specutils/blob/main/specutils/spectra/spectrum1d.py#L277-L284
although perhaps the coverage there is incomplete, don't know.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:265]: Yes, looking at the Mamba format I don't see any WCS info in MASK or IVAR; was just wondering what might happen when something is reading those HDUs separately, finding celestial axes without any coordinates.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:279]: Good point. There are at least 3 commonly used designations for StdDev, and we probably should also expect things like 'VARIAN' or 'INVVAR' for the others, so if anyone insists they have always used naming convention X in their project and it needs to stay that way, we should not prohibit that. Ultimately should be the responsibility of the user to use sensible values there _if_ they override the defaults, but I should probably add a check that at least it does not conflict with `UNCERT_REF`.

File: specutils/tests/test_loaders.py
Comment (dhomeier) [reply specutils/tests/test_loaders.py:797]: 😺 Maybe `5 * [0] + [1]` would be cooler still, but I did not really care as long as it fits on one line.
Do you mean is it necessary to use random numbers at all – probably not, just wanted to make up some more "realistic" example? Or to use `Generator` – that is how the `np.random` docs recommend it now.

**Comment by dhomeier** [comment]
Thanks for the review!

> Also, does this page need updating? It is not very clear to me exactly which data will go through this reader/writer.

It's the same `wcs1d-fits`, and ideally should handle all WCS (ImageHDU) spectra that do not have a more specialised reader. Auto-identify will likely not work (yet) e.g. for WCS with non-spectral components, but I think that column generally is not meant to say that it works for _all_ cases.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:226]: Actually it seems none of the functions in `default_loaders` have any docs built at all, though it would certainly be useful to have them and link them in https://specutils.readthedocs.io/en/latest/spectrum1d.html#list-of-loaders.

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:289]: Added to tests now, but nonstandard names do complicate everything quite a bit. Wondering if the uncertainty should actually written with the `BUNIT` – examples like MaNGA don't, but it would help with identifying the correct type (and also take care of the case that it might not match the flux unit in the file).

**Comment by pllim** [comment]
Sorry, I think my fixes to the tests caused a conflict here, so you need to rebase and resolve the conflict.

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [reply specutils/io/default_loaders/wcs_fits.py:289]: Yeah, trying to appease all the random file formats out there is going to be impossible.

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [inline specutils/io/default_loaders/wcs_fits.py:486]:
```suggestion
    Tuple of data to pass to `~specutils.SpectrumCollection` or `~specutils.Spectrum1D`:
```

Comment (pllim) [reply specutils/io/default_loaders/wcs_fits.py:486]: p.s. Does not really matter because this is a hidden function, so not rendered by API doc. But might as well do it right?

File: specutils/tests/test_loaders.py
Comment (pllim) [inline specutils/tests/test_loaders.py:983]:
```suggestion
    with fits.open(tmpfile, mode='update') as hdulist:
        hdulist[0].header.update([('PC1_2', 0.2)])
```

File: specutils/tests/test_loaders.py
Comment (pllim) [inline specutils/tests/test_loaders.py:1029]: Side note: Why call `os.system` instead of `shutil`? Anyways, I guess fixing this is out of scope.

**Review by pllim — COMMENTED** [review]
I left what general review comments I could. Now it is up to specutils maintainers to review. Thanks!

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:486]: Yes (none of the default_loaders are rendered anywhere afaics).

File: specutils/io/default_loaders/wcs_fits.py
Comment (rosteen) [inline specutils/io/default_loaders/wcs_fits.py:246]: I'm not _that_ familiar with the workings of WCS (and especially `wcs.to_fits()`) - is there any realistic scenario for setting `hdu` to a value other than 0, given that the initial `hdulist` is always created from the spectrum's WCS anyway? I guess what I'm asking is, is there really utility in having empty HDUs in front of the specified `hdu` index? There are no options to specify the HDU numbers for uncertainty and mask, so it's not like the user could write out in order [uncertainty, mask, flux], for example. Maybe we just get rid of the `hdu` argument altogether?

File: specutils/io/default_loaders/wcs_fits.py
Comment (rosteen) [inline specutils/io/default_loaders/wcs_fits.py:121]:
```suggestion
                                      f"will read '{hdu}'. Please use `hdu=<flux_hdu>` "
                                      "to select a specific extension.", AstropyUserWarning)
```
Minor quibble, would rather be specific ("extension" or "hdu") here than say "one".

File: specutils/io/default_loaders/wcs_fits.py
Comment (rosteen) [reply specutils/io/default_loaders/wcs_fits.py:289]: Agreed, we can only accommodate so much.

File: specutils/tests/test_loaders.py
Comment (rosteen) [reply specutils/tests/test_loaders.py:983]: +1 to this suggestion.

**Comment by rosteen** [comment]
Thanks @dhomeier, this is a great improvement. I had one question and left one suggestion, and there's one suggestion from @pllim remaining, but other than that I think this looks good. I appreciate the test coverage. 

I was going to say, maybe we change the name from `wcs1d_fits` to just `wcs_fits` if it handles multi-D WCS now, but perhaps we go with the standard answer of "1D means one spectral dimension" like in `Spectrum1D`.

File: specutils/io/default_loaders/wcs_fits.py
Comment (pllim) [reply specutils/io/default_loaders/wcs_fits.py:246]: As long as this is not considered "public" API and ok to break just like that, I agree.

File: specutils/tests/test_loaders.py
Comment (pllim) [reply specutils/tests/test_loaders.py:1029]: I don't see how "mv" cares about compression? https://docs.python.org/3/library/shutil.html#shutil.move

File: specutils/io/default_loaders/wcs_fits.py
Comment (dhomeier) [reply specutils/io/default_loaders/wcs_fits.py:246]: > I'm not _that_ familiar with the workings of WCS (and especially `wcs.to_fits()`) - is there any realistic scenario for setting `hdu` to a value other than 0, given that the initial `hdulist` is always created from the spectrum's WCS anyway?

It's not as much associated to using a WCS, but rather to the flux being stored in an ImageHDU, which can be the PrimaryHDU or an extension, while tables can only be stored in extensions. However for symmetry reasons or whatever sometimes writing the image to an extension, too, and using just a basic PrimaryHDU, is desired (there was one comment somewhere sketching just this scenario, but I cannot find it either in the Slack discussion or any related issue now).
I think originally accepting any value for `hdu` was just introduced as a poor substitute for not being able to write multiple spectra to one file/hdulist, but there is probably no real use case for this. For `hdu=1` I'd say there is, and this should should still remain available.

File: specutils/tests/test_loaders.py
Comment (dhomeier) [reply specutils/tests/test_loaders.py:1029]: Not at all, I guess, but it's kind of the direct counterpart to `os.system(f'{compress} {tmpfile}')` just above.

File: specutils/io/default_loaders/wcs_fits.py
Comment (rosteen) [reply specutils/io/default_loaders/wcs_fits.py:246]: Thanks for the explanation...I guess it makes sense to leave it in then for now.

**Comment by rosteen** [comment]
@dhomeier Can I commit the last suggestions for you and merge this?

**Review by rosteen — APPROVED** [review]
Approved, need to commit last two suggestions before merging.

**Comment by pllim** [comment]
Hmm does this mean we should ditch Jdaviz custom writer and use this?

**Comment by rosteen** [comment]
> > @dhomeier Can I commit the last suggestions for you and merge this?
> 
> I've included them in some more local updates I had made; should be good to merge then.

Blast, I went ahead and committed and merged, thinking you might be done and logged off for the day. Are your other updates worth opening a follow-up PR for? Sorry about that.

**Comment by dhomeier** [comment]
Ah, remote tests are taking too long here (spPlate really bogs everything down).
Will think about the follow-up, the extra test might be good to have, but now I am going to be offline for a while.

**Comment by pllim** [comment]
Thanks, @dhomeier ! I am trying out this improved format over at https://github.com/spacetelescope/jdaviz/pull/2094 but I encountered error. Perhaps we need an extra step to convert bool mask to uint16 (I guess it could be uint8 but uint16 is common in bit-plane format)?

```
>       fitted_spectrum.write(out_fn, format="wcs1d-fits", overwrite=True)

jdaviz/configs/default/plugins/model_fitting/tests/test_fitting.py:229: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/nddata/mixins/ndio.py:103: in __call__
    self.registry.write(self._instance, *args, **kwargs)
astropy/io/registry/core.py:383: in write
    return writer(data, *args, **kwargs)
specutils/io/default_loaders/wcs_fits.py:278: in wcs1d_fits_writer
    hdulist[-1].data = spectrum.mask
astropy/utils/decorators.py:848: in __set__
    ret = self.fset(obj, val)
...
E           KeyError: 'bool'
```

https://github.com/spacetelescope/jdaviz/actions/runs/4451633462/jobs/7818514722?pr=2094

**Pull request merged**
